### PR TITLE
Allowing traffic from different k8s namespaces

### LINKED ISSUE #966: Inter-namespace CiliumNetworkPolicy
Right now CiliumNetworkPolicy gets created inside a kubernetes namespace, and you cannot override source and destination endpoint. It would be nice to be able to create a policy where source and destination are in different namespaces. Ideally I'd like to be able to do something like this:

```
kind: CiliumNetworkPolicy
apiVersion: cilium.io/v1
metadata:
  name: my-policy
spec:
  endpointSelector:
    matchLabels:
      "k8s:io.kubernetes.pod.namespace": "n1"
  ingress:
    - fromEndpoints:
        - matchLabels:
          "k8s:io.kubernetes.pod.namespace": "n2"
      toPorts:
        - ports:
            - port: '80'
              protocol: TCP
          rules: {}
```

### PR DESCRIPTION
**Pull request by aanm**
Fixes #966

### CODE REVIEW

File: pkg/k8s/third_party.go [hunk 1]
```diff
@@ -82,7 +82,13 @@ func (r *CiliumNetworkPolicy) Parse() (api.Rules, error) {
 					if retRule.Ingress[i].FromEndpoints[j].HasKeyPrefix(labels.LabelSourceReservedKeyPrefix) {
 						continue
 					}
-					retRule.Ingress[i].FromEndpoints[j].MatchLabels[labels.LabelSourceK8sKeyPrefix+PodNamespaceLabel] = namespace
+					// Store the same namespace on the "FromEndpoints" map
```
Comment (tgraf) [inline pkg/k8s/third_party.go]: Let's reword this comment:
The user can explicitly specify the namespace in the FromEndpoints selector. If omitted, we limit the scope to the namespace the policy lives in.

**Review by tgraf — CHANGES_REQUESTED** [review]
Looks good. Can you add a warning in this case:
```
retRule.EndpointSelector.LabelSelector.MatchLabels[labels.LabelSourceK8sKeyPrefix+PodNamespaceLabel] = namespace
```
If namespace label is already set, and not identical to namespace in metadata, then let's print a warning for now.

File: pkg/k8s/third_party.go [hunk 2]
```diff
@@ -66,7 +68,14 @@ func (r *CiliumNetworkPolicy) Parse() (api.Rules, error) {
...
 			retRule.EndpointSelector.LabelSelector.MatchLabels = map[string]string{}
 		}
-		retRule.EndpointSelector.LabelSelector.MatchLabels[labels.LabelSourceK8sKeyPrefix+PodNamespaceLabel] = namespace
+
+		userNamespace, ok := retRule.EndpointSelector.LabelSelector.MatchLabels[labels.LabelSourceK8sKeyPrefix+PodNamespaceLabel]
+		if ok && userNamespace != namespace {
+			log.Warningf("k8s: CiliumNetworkPolicy %s/%s contains illegal namespace match '%s' in EndpointSelector."+
+				" EndpointSelector always applies in namespace of the policy resource, removing namespace match '%s'.",
+				r.Metadata.Namespace, r.Metadata.Name, userNamespace, userNamespace)
+			retRule.EndpointSelector.LabelSelector.MatchLabels[labels.LabelSourceK8sKeyPrefix+PodNamespaceLabel] = namespace
```
Comment (tgraf) [inline pkg/k8s/third_party.go]: This looks wrong. You always want to write the label but omit a warning if there was a mismatch. The assignment should be outside of the conditional.

**Review by tgraf — CHANGES_REQUESTED** [review]

File: pkg/k8s/third_party.go (hunk 2 above)
Comment (aanm) [reply pkg/k8s/third_party.go]: Fixed. At least the unit tests also failed on jenkins because of this

File: tests/helpers.bash [hunk 3]
```diff
@@ -139,6 +139,22 @@ function wait_for_policy_enforcement {
     done
 }
 
+function wait_for_cilium_k8s_ep_gen {
+    export cilium_k8s_npods=$(kubectl get pods -n kube-system | grep cilium | wc -l)
```
Comment (tgraf) [inline tests/helpers.bash]: Why does this need to be exported? This should be prefixed with `local`

File: tests/helpers.bash [hunk 4]
```diff
@@ -139,6 +139,22 @@ function wait_for_policy_enforcement {
     done
 }
 
+function wait_for_cilium_k8s_ep_gen {
```
Comment (tgraf) [inline tests/helpers.bash]: Please give this a proper name: `wait_all_k8s_regenerated`

File: tests/helpers.bash [hunk 5]
```diff
@@ -139,6 +139,22 @@ function wait_for_policy_enforcement {
...
+    export cilium_k8s_npods=$(kubectl get pods -n kube-system | grep cilium | wc -l)
+    set +x
+    { for i in $(seq 1 ${cilium_k8s_npods}); do
+        while true; do
+            cilium_pod_id=$(kubectl get pods -n kube-system | grep cilium | awk "NR==$i{ print \$1 }")
+            if ! kubectl exec -n kube-system -i ${cilium_pod_id} cilium endpoint list | grep regenerating; then
```
Comment (tgraf) [inline tests/helpers.bash]: I don't think this is safe. This assumes that by the name you call this, at least endpoint has already been marked to be in regenerating state. That might not be true, regeneration might be delayed to after the first iteration of this loop has been executed.

**Review by tgraf — CHANGES_REQUESTED** [review]

File: tests/helpers.bash (hunk 5 above)
Comment (tgraf) [reply tests/helpers.bash]: I'm fine deferring this to later but it needs a fat FIXME comment

Comment (aanm) [reply tests/helpers.bash]: I've added a FIXME and added a FIXME on the function above since it might have the same problem.

**Review by tgraf — APPROVED** [review]

**Review by aalemayhu — COMMENTED** [review]
LGTM.

**Pull request merged**